clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/agent/CleanupAlarmsAgent.groovy

```diff
@@ -86,36 +86,38 @@ class CleanupAlarmsAgent implements RunnableAgent, CustomScheduledAgent {
       credentials.regions.each { AmazonCredentials.AWSRegion region ->
         log.info("Looking for alarms to delete")
-
-        def cloudWatch = amazonClientProvider.getCloudWatch(credentials, region.name)
-        Set<String> attachedAlarms = getAttachedAlarms(amazonClientProvider.getAutoScaling(credentials, region.name))
-        def describeAlarmsRequest = new DescribeAlarmsRequest().withStateValue(StateValue.INSUFFICIENT_DATA)
-
-        while (true) {
-          def result = cloudWatch.describeAlarms(describeAlarmsRequest)
-
-          List<MetricAlarm> alarmsToDelete = result.metricAlarms.findAll {
-            it.stateUpdatedTimestamp.before(DateTime.now().minusDays(daysToLeave).toDate()) &&
-              !attachedAlarms.contains(it.alarmName) &&
-              ALARM_NAME_PATTERN.matcher(it.alarmName).matches()
-          }
-
-          if (alarmsToDelete) {
-            // terminate up to 20 alarms at a time (avoids any AWS limits on # of concurrent deletes)
-            alarmsToDelete.collate(20).each {
-              log.info("Deleting ${it.size()} alarms in ${credentials.name}/${region.name} " +
-                "(alarms: ${it.alarmName.join(", ")})")
-              cloudWatch.deleteAlarms(new DeleteAlarmsRequest().withAlarmNames(it.alarmName))
-              Thread.sleep(500)
+        try {
+          def cloudWatch = amazonClientProvider.getCloudWatch(credentials, region.name)
+          Set<String> attachedAlarms = getAttachedAlarms(amazonClientProvider.getAutoScaling(credentials, region.name))
+          def describeAlarmsRequest = new DescribeAlarmsRequest().withStateValue(StateValue.INSUFFICIENT_DATA)
+
+          while (true) {
+            def result = cloudWatch.describeAlarms(describeAlarmsRequest)
+
+            List<MetricAlarm> alarmsToDelete = result.metricAlarms.findAll {
+              it.stateUpdatedTimestamp.before(DateTime.now().minusDays(daysToLeave).toDate()) &&
+                !attachedAlarms.contains(it.alarmName) &&
+                ALARM_NAME_PATTERN.matcher(it.alarmName).matches()
             }
 
-          }
+            if (alarmsToDelete) {
+              // terminate up to 20 alarms at a time (avoids any AWS limits on # of concurrent deletes)
+              alarmsToDelete.collate(20).each {
+                log.info("Deleting ${it.size()} alarms in ${credentials.name}/${region.name} " +
+                  "(alarms: ${it.alarmName.join(", ")})")
+                cloudWatch.deleteAlarms(new DeleteAlarmsRequest().withAlarmNames(it.alarmName))
+                Thread.sleep(500)
+              }
+            }
 
-          if (result.nextToken) {
-            describeAlarmsRequest.withNextToken(result.nextToken)
-          } else {
-            break
+            if (result.nextToken) {
+              describeAlarmsRequest.withNextToken(result.nextToken)
+            } else {
+              break
+            }
           }
+        } catch (Exception e) {
```

Comment:
Is there a catch block to add "higher up" (where the run method is called) that would (also) help?  Like, what if there exceptions in other agents?

Reply:
Higher up there is already a try/catch block in the `RunnableAgent ` initial implementation  but i have pushed a try/catch block in the `CleanupDetachedInstancesAgent` as well which implements a RunnableAgent

Reply:
Seems like `run` is called from [RunnableAgentExecution.executeAgent](https://github.com/spinnaker/clouddriver/blob/f5f4b112f7b8a5d97127b91b859277fab99a6a59/cats/cats-core/src/main/java/com/netflix/spinnaker/cats/agent/RunnableAgent.java#L32) and all the places that call `executeAgent` are similar to [DefaultAgentScheduler](https://github.com/spinnaker/clouddriver/blob/f5f4b112f7b8a5d97127b91b859277fab99a6a59/cats/cats-core/src/main/java/com/netflix/spinnaker/cats/agent/DefaultAgentScheduler.java#L134).  There's a try/catch, but it only updates metrics.  So what we get here is some extra logging, but I don't see how it's going to help with agent scheduling.

Reply:
This cleanup agents are going through each account and trying to do a cleanup of any stale alarms in cloudwatch or any detached instances etc. When a bad account is present in the credentials repository the agent stops and doesnt go through the rest of the accounts.
By adding a try/catch here we are logging that there was a problem with X account and continue to the next account.

Reply:
Aaah yes, the try/catch here and in CleanupDetachedInstancesAgent is inside `getAccounts().each`.  Seems like the consequences of this fix are good...that instead of dying on the first error, we'll continue to clean up other accounts.